topology-manager/src/main/java/org/apache/aries/rsa/topologymanager/importer/MultiMap.java

```diff
@@ -19,60 +19,54 @@
 package org.apache.aries.rsa.topologymanager.importer;
 
 import java.util.Collections;
-import java.util.HashMap;
-import java.util.HashSet;
 import java.util.Map;
 import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.CopyOnWriteArraySet;
 
 /**
- * Minimal implementation of a synchronized map
+ * Minimal implementation of a thread-safe map where each key can have multiple values.
+ *
+ * @param <K> the key type
+ * @param <V> the value type
  */
-public class MultiMap<T> {
+public class MultiMap<K, V> {
 
-    private Map<String, Set<T>> map;
+    private Map<K, Set<V>> map;
     
     public MultiMap() {
-        map = new HashMap<>();
-    }
-    
-    public synchronized void put(String key, T value) {
-        Set<T> values = map.get(key);
-        if (values == null) {
-            values = new HashSet<>();
-            map.put(key, values);
-        }
-        values.add(value);
+        map = new ConcurrentHashMap<>();
     }
     
-    public synchronized Set<T> get(String key) {
-        if (map.containsKey(key)) {
-            return Collections.unmodifiableSet(new HashSet<>(map.get(key)));
-        } else {
-            return Collections.<T>emptySet();
-        }
+    public void put(K key, V value) {
+        map.compute(key, (k, v) -> {
+            if (v == null)
+                v = new CopyOnWriteArraySet<>();
+            v.add(value);
+            return v;
+        });
     }
 
-    public synchronized void remove(String key, T value) {
-        Set<T> values = map.get(key);
-        values.remove(value);
-        if (values.isEmpty()) {
-            map.remove(key);
-        }
+    // the returned set should not be modified directly
+    public Set<V> get(K key) {
+        return map.getOrDefault(key, Collections.emptySet());
```

Comment:
As values should only be added/removed through this MultiMap implementation, I would propose to wrap the result of getOrDefault with Collection.unmodifiableSet()

Reply:
Sure. Even better, I'll wrap it only if it's not the empty set.